src/Workspaces/Core/Portable/Serialization/IsolatedAnalyzerReference.cs

```diff
@@ -0,0 +1,106 @@
+/// references.  This allows for loading the analyzers and generators from it in a way that is associated with that load
+/// context, keeping them separate from other analyzers and generators loaded in other load contexts, while also
+/// allowing all of those instances to be collected when no longer needed.  Being isolated means that if any of the
+/// underlying assembly references change, that they can be loaded side by side with the prior references.  This enables
+/// functionality like live reloading of analyzers and generators when they change on disk.  Note: this is only
+/// supported on .Net Core, and not .Net Framework, as only the former has <see cref="AssemblyLoadContext"/>s.
+/// </summary>
+/// <remarks>
+/// The purpose of this type is to allow passing out a <see cref="AnalyzerReference"/> to the rest of the system that
+/// then ensures that as long as it is alive (or any <see cref="DiagnosticAnalyzer"/> or <see cref="ISourceGenerator"/>
+/// it passes out is alive), that the <see cref="IsolatedAssemblyReferenceSet"/> (and its corresponding <see
+/// cref="AssemblyLoadContext"/>) is kept alive as well.
+/// </remarks>
+internal sealed class IsolatedAnalyzerReference(
+    IsolatedAssemblyReferenceSet isolatedAssemblyReferenceSet,
+    AnalyzerReference underlyingAnalyzerReference) : AnalyzerReference
+{
+    /// <summary>
+    /// Conditional weak tables that ensure that as long as a particular <see cref="DiagnosticAnalyzer"/> or <see
+    /// cref="ISourceGenerator"/> is alive, that the corresponding <see cref="IsolatedAssemblyReferenceSet"/> (and its
+    /// corresponding <see cref="AssemblyLoadContext"/> is kept alive.
+    /// </summary>
+    private static readonly ConditionalWeakTable<DiagnosticAnalyzer, IsolatedAssemblyReferenceSet> s_analyzerToPinnedReferenceSet = [];
+
+    /// <inheritdoc cref="s_analyzerToPinnedReferenceSet"/>
+    private static readonly ConditionalWeakTable<ISourceGenerator, IsolatedAssemblyReferenceSet> s_generatorToPinnedReferenceSet = [];
+
+    /// <summary>
+    /// We keep a strong reference here.  As long as this <see cref="IsolatedAnalyzerReference"/> is passed out and held
+    /// onto (say by a Project instance), it should keep the IsolatedAssemblyReferenceSet (and its ALC) alive.
+    /// </summary>
+    private readonly IsolatedAssemblyReferenceSet _isolatedAssemblyReferenceSet = isolatedAssemblyReferenceSet;
+
+    /// <summary>
+    /// The actual real <see cref="AnalyzerReference"/> we defer our operations to.
+    /// </summary>
+    public readonly AnalyzerReference UnderlyingAnalyzerReference = underlyingAnalyzerReference;
+
+    public override string Display => UnderlyingAnalyzerReference.Display;
+    public override string? FullPath => UnderlyingAnalyzerReference.FullPath;
+    public override object Id => UnderlyingAnalyzerReference.Id;
+
+    public override ImmutableArray<DiagnosticAnalyzer> GetAnalyzers(string language)
```

Comment:
You _may_ wish to consider putting an explicit GC.KeepAlive(this) or GC.KeepAlive(_isolatedAssemblyReferenceSet) for sanity here. I was mentally confirming that there's not a tiny window of time where the a call to here fetches the UnderlyingAnalyzerReference, and calls GetAnalyzers on it, and when the object is returned _but before we register it in the ConditionalWeakTable_, we end up with the set GCing. It's fine since the later call to PinAnalyzers still needs it. But it's still subtle.

Reply:
wow.  good call.

Reply:
ok.  i actually think this is NOT possible.   The reason for this is: 

>  and when the object is returned but before we register it in the ConditionalWeakTable, 

In order to register with CWT we necessarily are then holding ont the IsolatedSet (it's what goes into the CWT after all).  So it's impossible for it to get GCed prior to that point.  

That said, i changed this up a bit.  please take a look.